### ISSUE TITLE
`default-font-size` and `default-font-family` don't work in the Live-Preview

### ISSUE THREAD

**Issue by asuper0** [slint-ui/slint#4298]
System: Win10
VS code :1.83.0
Slint plugin:1.3.2

The `default-font-size` property just don't work in preview, but work when build and run.

**Comment by tronical** [comment]
Yes indeed, this also applies to the other `default-font` properties. When resolved at run-time, they resolve against the `Window` of the preview itself, not the `Window` in the user's Slint code :-(

I'm not sure how to solve that best yet...

**Comment by tronical** [comment]
We could change the signatures of say `fn layout_info(..., window_adapter: &Rc<dyn WindowAdapter>)` to something like `fn layout_info(..., window_context: &WindowContext)` and let `WindowContext` be a struct with a ref to the window adapter as well as the `Pin<&WindowItem>`. But besides the difficulty of implementing that, it also wouldn't help for `render()` itself, when we have to do the same resolution step inside the specific renderer and certainly don't know which `WindowItem` to pick.

Unless......... perhaps the other way of doing this is - at resolution time - to walk /up/ the item tree, starting at `self_rc` and then `parent()` until... we find the `Window`. That's not great, but it would be correct.

**Comment by ogoffart** [comment]
Another slightly related issue is that if one use `default-font-size`, it will not work when previewing sub components.
Perhaps default-font-size should be a "project" property (#267)?  But we would still need to solve the problem on where to get the default font.

I guess the short term solution is indeed to walk up the item tree. But that is a fairly slow operation, we might not want to do that for every rendering of every Text.  Adding some kind of WindowContext might not be a bad idea either.

**Comment by tronical** [comment]
OIivier had a great idea for an optimisation here: We have to walk up the tree, but instead of traversing item by item we can traverse from component to component and just check if item at index 0 is a WindowItem. If it isn't, continue the walk.

**Comment by ubruhin** [comment]
I'd like to vote for this feature too. Sometimes it's quite cumbersome to tune a UI with the live preview just to realize afterwards that the compiled application looks very different (it's not only the different text rendering as the effect of different text heights can propagate through layouts also to many other components).

At least I suspect the different text heights are caused by this issue.

**Comment by igiona** [comment]
I agree that this one degrades the development cycle more than one might think at a first glance.

As SW developer, I can easily build & run the app.
But a designer or marketing folks don't really want that.
The preview is amazing for that, a huge plus for Slint.
With this bug in place though, what you see is not what you get, and for designers that matters!

**Comment by tronical** [comment]
Ooops, I accidentally closed this, but there's one missing bit: `rem` resolution is missing here. It uses `BuiltinFunction::GetWindowDefaultFontSize`, which won't traverse from the "current" item but goes straight to the next window adapter.
